Notify/handler.ts

```diff
@@ -8,24 +13,190 @@ import { RequiredBodyPayloadMiddleware } from "@pagopa/io-functions-commons/dist
   IResponseErrorForbiddenNotAuthorized,
-  IResponseSuccessNoContent
+  IResponseErrorInternal,
+  IResponseErrorNotFound,
+  IResponseSuccessNoContent,
+  ResponseErrorInternal,
+  ResponseSuccessNoContent
 } from "@pagopa/ts-commons/lib/responses";
 import { initAppInsights } from "@pagopa/ts-commons/lib/appinsights";
+import { FiscalCode, NonEmptyString } from "@pagopa/ts-commons/lib/strings";
 import { NotificationInfo } from "../generated/definitions/NotificationInfo";
+import {
+  NotificationType,
+  NotificationTypeEnum
+} from "../generated/definitions/NotificationType";
+
+import {
+  getPrinterForTemplate,
+  NotificationPrinter
+} from "../templates/printer";
+
+import { SendNotification } from "./notification";
+import {
+  MessageWithContentReader,
+  ServiceReader,
+  SessionStatusReader
+} from "./readers";
+
+const isReminderNotification = (notificationType: NotificationType): boolean =>
+  [
+    NotificationTypeEnum.REMINDER_PAYMENT,
+    NotificationTypeEnum.REMINDER_PAYMENT_LAST,
+    NotificationTypeEnum.REMINDER_READ
+  ].includes(notificationType);
 
-type NotifyHandler = () => Promise<
-  IResponseSuccessNoContent | IResponseErrorForbiddenNotAuthorized
+/**
+ * Check whether a Reminder notification can be sent to user or not
+ * NOTE: Right now we got no opt-in check. Always return true
+ *
+ * @param fiscalCode The user Fiscal Code
+ * @returns
+ */
+const canSendReminderNotification = (
+  _fiscalCode: FiscalCode
+): TE.TaskEither<Error, boolean> => TE.right(true);
+
+/**
+ * Check whether a notification can be sent to user
+ *
+ * @param notificationType
+ * @param fiscalCode
+ * @returns a TaskEither of Error or boolean
+ */
+const checkSendNotificationPermission = (
+  notificationType: NotificationType,
+  fiscalCode: FiscalCode
+): TE.TaskEither<Error, boolean> =>
+  match(notificationType)
+    .when(isReminderNotification, _ => canSendReminderNotification(fiscalCode))
+    // Not implemented yet
+    .otherwise(_ => TE.of(false));
+
+/**
+ * Check whether a notification should not contain personal information
+ * depending on user' session status
+ *
+ * @param notificationType
+ * @param fiscalCode
+ * @returns a TaskEither of Error or boolean
+ */
+const canSendVerboseNotification = (
+  retrieveUserSession: SessionStatusReader,
+  fiscalCode: FiscalCode
+): TE.TaskEither<Error, boolean> =>
+  pipe(
+    retrieveUserSession(fiscalCode),
+    TE.mapLeft(_ => Error("Error checking user session")),
+    TE.map(sessionStatus => sessionStatus.active)
+  );
+
+const prepareNotification = (
+  retrieveUserSession: SessionStatusReader,
+  retrieveMessageWithContent: MessageWithContentReader,
+  retrieveService: ServiceReader
+) => (
+  notification_type: NotificationTypeEnum,
+  fiscal_code,
+  message_id
+): TE.TaskEither<
+  IResponseErrorInternal | IResponseErrorNotFound,
+  NotificationPrinter
+> =>
+  pipe(
+    canSendVerboseNotification(retrieveUserSession, fiscal_code),
+    TE.orElse(_err => TE.of(false)),
```

Comment:
Are we sure we want to consider an error with a default behaviour? (i.e: BE not reacheable so we send an anonymous notification)

Reply:
This is an open point, you're right.
Is it better to send them anyway anonymously, or is it better to block all notifications till the error is gone?
I don't have a strong opinion on that.